kernel/src/sync/mqueue.rs

```diff
@@ -0,0 +1,625 @@
+            node_count,
+            accessible_count: [AtomicU32::new(node_count), AtomicU32::new(0)],
+            pend_queues: [const { SpinLock::new(WaitQueue::new()) }; 2],
+            queue_buffer: if buf.is_null() {
+                BoxedRingBuffer::new(total_size)
+            } else {
+                BoxedRingBuffer::new_with_mem(total_size, buf)
+            },
+            lock: ISpinLock::new(),
+        }
+    }
+
+    pub fn init(&self) -> bool {
+        self.pend_queues[SEND_TYPE].irqsave_lock().init()
+            && self.pend_queues[RECV_TYPE].irqsave_lock().init()
+    }
+
+    #[inline]
+    pub fn lock(&self) -> SpinLockGuard<'_, Self> {
+        self.lock.irqsave_lock()
+    }
+
+    #[inline]
+    pub fn node_info(&self) -> (usize, u32) {
+        (self.node_size, self.node_count)
+    }
+
+    #[inline]
+    pub fn sendable_count(&self) -> u32 {
+        self.accessible_count[SEND_TYPE].load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    pub fn increment_sendable_count(&self) {
+        self.accessible_count[SEND_TYPE].fetch_add(1, Ordering::Relaxed);
+    }
+
+    #[inline]
+    pub fn decrement_sendable_count(&self) {
+        self.accessible_count[SEND_TYPE].fetch_sub(1, Ordering::Relaxed);
+    }
+
+    #[inline]
+    pub fn recvable_count(&self) -> u32 {
+        self.accessible_count[RECV_TYPE].load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    pub fn increment_recvable_count(&self) {
+        self.accessible_count[RECV_TYPE].fetch_add(1, Ordering::Relaxed);
+    }
+
+    #[inline]
+    pub fn decrement_recvable_count(&self) {
+        self.accessible_count[RECV_TYPE].fetch_sub(1, Ordering::Relaxed);
+    }
+
+    fn wakeup_pend_receiver(w: &mut SpinLockGuard<'_, WaitQueue>) -> bool {
+        if !w.is_empty() {
+            let next = w.pop_front().unwrap();
+            let t = next.thread.clone();
+            if let Some(timer) = &t.timer {
+                timer.stop();
+            }
+            let ok = scheduler::queue_ready_thread(thread::SUSPENDED, t.clone());
+            assert!(ok);
+            return true;
+        }
+        false
+    }
+
+    pub fn send(
+        &self,
+        buffer: &[u8],
+        size: usize,
+        timeout: usize,
+        urgent: SendMode,
+    ) -> Result<(), Error> {
+        if buffer.len() < size {
+            return Err(code::EINVAL);
+        }
+        let head_size = core::mem::size_of::<usize>();
+        let mut timeout = timeout;
+
+        if size > (self.node_size - head_size) {
+            return Err(code::EOVERFLOW);
+        }
+
+        let mut queue = self.lock();
+        while self.sendable_count() == 0 {
+            if timeout == NO_WAITING {
+                return Err(code::ETIMEDOUT);
+            }
+            if irq::is_in_irq() {
+                return Err(code::ENOTSUP);
+            }
+            let mut ticks = time::get_sys_ticks();
+            let mut sendqueue = self.pend_queues[SEND_TYPE].irqsave_lock();

```

Comment:
```suggestion
            let mut send_queue = self.pend_queues[SEND_TYPE].irqsave_lock();
```
please keep consistent naming style.